src/wp-includes/https-detection.php

```diff
@@ -0,0 +1,139 @@
+<?php
+/**
+ * HTTPS detection functions.
+ *
+ * @package WordPress
+ * @since 5.7.0
+ */
+
+/**
+ * Checks whether the website is using HTTPS.
+ *
+ * This is based on whether the home and site URL are using HTTPS.
+ *
+ * @since 5.7.0
+ *
+ * @return bool True if using HTTPS, false otherwise.
+ */
+function wp_is_using_https() {
+	if ( 'https' !== wp_parse_url( home_url(), PHP_URL_SCHEME ) ) {
+		return false;
+	}
+
+	// Use direct option access for 'siteurl' and manually run the 'site_url'
+	// filter because site_url() will adjust the scheme based on what the
+	// current request is using.
+	/** This filter is documented in wp-includes/link-template.php */
+	$site_url = apply_filters( 'site_url', get_option( 'siteurl' ), '', null, null );
+
+	if ( 'https' !== wp_parse_url( $site_url, PHP_URL_SCHEME ) ) {
+		return false;
+	}
+
+	return true;
+}
+
+/**
+ * Checks whether HTTPS is supported for the server and domain.
+ *
+ * @since 5.7.0
+ *
+ * @return bool True if HTTPS is supported, false otherwise.
+ */
+function wp_is_https_supported() {
+	$https_detection_errors = get_option( 'https_detection_errors' );
+
+	// If option has never been set by the Cron hook before, run it on-the-fly as fallback.
+	if ( false === $https_detection_errors ) {
+		wp_update_https_detection_errors();
+
+		$https_detection_errors = get_option( 'https_detection_errors' );
+	}
+
+	// If there are no detection errors, HTTPS is supported.
+	return empty( $https_detection_errors );
+}
+
+/**
+ * Runs a remote HTTPS request to detect whether HTTPS supported, and stores potential errors.
+ *
+ * This internal function is called by a regular Cron hook to ensure HTTPS support is detected and maintained.
+ *
+ * @since 5.7.0
+ * @access private
+ */
+function wp_update_https_detection_errors() {
+	$support_errors = new WP_Error();
+
+	$response = wp_remote_request(
+		home_url( '/', 'https' ),
```

Comment:
As opposed to getting the `home_url( '/' )`, I wonder if it would be better to hit the REST API, perhaps to fetch `wp-json/wp/v2/settings` and then examine that the returned home URL and site URL match what is expected.

I'm thinking that this could address a scenario where a web server is misconfigured to serve two separate sites over HTTPS vs HTTP, or that accessing the site via HTTPS could return some Nginx welcome screen. If that happened, then the code currently will have a false positive that HTTPS is supported. If, however, an authenticated HTTPS REST API request is made for the settings, then this would guarantee that the site is indeed alive at that URL.

The PWA plugin addresses this by verifying that the expected Web App Manifest is in the HTML response: https://github.com/GoogleChromeLabs/pwa-wp/blob/17e23ef28e9224c5cdbc839596377e10c7b158a2/wp-includes/class-wp-https-detection.php#L133-L139

Reply:
@westonruter I think calling the REST API for this would come with its own issues, since some sites generally block it or hide it behind authentication etc. But it's a good point we should make this more reliable. I like the idea of checking for something WordPress site-specific in the HTML output, but I'm wondering what we could rely on. Can you think of something appropriate that should really always be there and developers/plugins likely won't (or even better can't) remove it?

Reply:
In the PWA plugin we relied on the `link[rel=manifest]` element but that's not yet available… 

The `wlwmanifest_link()` and `rsd_link()` functions are unconditionally added to `wp_head`, but site authors can always remove them. I suppose we could check if `has_action()` for either of those, and then look for them. The same goes for the RSS feed link. 

The check could see if any of these are present, and then look for them. Otherwise, if none of the actions are added, then it could just fall back to a successful HTTP request without verifying the response body.

Reply:
@westonruter I've added a new private function `wp_is_owned_html_output()` for that purpose. Let me know your thoughts.